File: 06__full-stack/brainrot-studio/backend/app/schemas/timeline.py
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.scene import TransitionType
from app.models.scene_asset import SceneAssetRole
from app.models.track import TrackType


class CompositionRead(BaseModel):
    id: UUID
    video_id: UUID
    width: int
    height: int
    fps: int
    duration_ms: int


class CompositionUpdate(BaseModel):
    width: int | None = Field(default=None, ge=100)
    height: int | None = Field(default=None, ge=100)
    fps: int | None = Field(default=None, ge=1)


class SceneAssetCreate(BaseModel):
    asset_id: UUID
    role: SceneAssetRole
    start_ms: int = 0
    duration_ms: int | None = None
    z_index: int = 0
    x: float = 0.5
    y: float = 0.5
    scale: float = 1.0
    rotation: float = 0.0
    opacity: float = 1.0


class SceneAssetUpdate(BaseModel):
    role: SceneAssetRole | None = None
    start_ms: int | None = None
    duration_ms: int | None = None
    z_index: int | None = None
    x: float | None = None
    y: float | None = None
    scale: float | None = None
    rotation: float | None = None
    opacity: float | None = None


class SceneAssetRead(BaseModel):
    id: UUID
    scene_id: UUID
    asset_id: UUID
    role: SceneAssetRole
    start_ms: int
    duration_ms: int | None
    z_index: int
    x: float
    y: float
    scale: float
    rotation: float
    opacity: float


class SceneReadWithAssets(BaseModel):
    id: UUID
    video_id: UUID
    order: int
    position: int
    start_ms: int
    duration_ms: int
    title: str | None
    description: str | None
    narration: str | None
    visual_prompt: str | None
    dialogue: str | None
    transition_in: TransitionType
    assets: list[SceneAssetRead] = []


class TrackItemRead(BaseModel):
    id: UUID
    track_id: UUID
    asset_id: UUID | None
    start_ms: int
    duration_ms: int
    offset_ms: int


class TrackItemCreate(BaseModel):
    asset_id: UUID | None = None
    start_ms: int
    duration_ms: int
    offset_ms: int = 0


class TrackRead(BaseModel):
    id: UUID
    video_id: UUID
    name: str
    track_type: TrackType
    order: int
    muted: bool
    items: list[TrackItemRead] = []


class TrackCreate(BaseModel):
    name: str
    track_type: TrackType
    order: int = 0
    muted: bool = False


class CaptionRead(BaseModel):
    id: UUID
    video_id: UUID
    text: str
    start_ms: int
    end_ms: int
    style: str


class CaptionCreate(BaseModel):
    text: str
    start_ms: int
    end_ms: int
    style: str = "default"


class TimelineRead(BaseModel):
    composition: CompositionRead
    scenes: list[SceneReadWithAssets]
    tracks: list[TrackRead]
    captions: list[CaptionRead]


class SceneReorderRequest(BaseModel):
    scene_ids: list[UUID]

File: 06__full-stack/brainrot-studio/frontend/src/lib/types/timeline.ts
export interface Composition {
	id: string;
	video_id: string;
	width: number;
	height: number;
	fps: number;
	duration_ms: number;
}

export type SceneAssetRole =
	| 'background'
	| 'character'
	| 'prop'
	| 'audio'
	| 'music'
	| 'voice'
	| 'overlay';

export interface SceneAsset {
	id: string;
	scene_id: string;
	asset_id: string;
	role: SceneAssetRole;
	start_ms: number;
	duration_ms: number | null;
	z_index: number;
	x: number;
	y: number;
	scale: number;
	rotation: number;
	opacity: number;
}

export type TransitionType = 'cut' | 'fade' | 'slide' | 'zoom';

export interface SceneWithAssets {
	id: string;
	video_id: string;
	order: number;
	position: number;
	start_ms: number;
	duration_ms: number;
	title: string | null;
	description: string | null;
	narration: string | null;
	visual_prompt: string | null;
	dialogue: string | null;
	transition_in: TransitionType;
	assets: SceneAsset[];
}

export type TrackType = 'video' | 'audio' | 'caption';

export interface TrackItem {
	id: string;
	track_id: string;
	asset_id: string | null;
	start_ms: number;
	duration_ms: number;
	offset_ms: number;
}

export interface Track {
	id: string;
	video_id: string;
	name: string;
	track_type: TrackType;
	order: number;
	muted: boolean;
	items: TrackItem[];
}

export interface Caption {
	id: string;
	video_id: string;
	text: string;
	start_ms: number;
	end_ms: number;
	style: string;
}

export interface Timeline {
	composition: Composition;
	scenes: SceneWithAssets[];
	tracks: Track[];
	captions: Caption[];
}
